clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CompletableBackgroundEvent.java

```diff
@@ -27,19 +30,34 @@
 public abstract class CompletableBackgroundEvent<T> extends BackgroundEvent implements CompletableEvent<T> {
 
     private final CompletableFuture<T> future;
+    private final long deadlineMs;
 
-    protected CompletableBackgroundEvent(final Type type) {
+    protected CompletableBackgroundEvent(final Type type, final Timer timer) {
         super(type);
         this.future = new CompletableFuture<>();
+        Objects.requireNonNull(timer);
+
+        long currentTimeMs = timer.currentTimeMs();
+        long remainingMs = timer.remainingMs();
+
+        if (currentTimeMs > Long.MAX_VALUE - remainingMs)
+            this.deadlineMs = Long.MAX_VALUE;
+        else
+            this.deadlineMs = currentTimeMs + remainingMs;
```

Comment:
Same questions as above.

Reply:
Per the above, I added `CompletableEvent.calculateDeadlineMs()` to keep the code in a shared location.